### DISCUSSION TITLE
How to extend zune-image?

### DISCUSSION THREAD

**Discussion by vlad-salone** [etemesi254/zune-image#161] [Q&A]
Hello, I am building my CLI image processing tool that leverages the usage of modern codecs such as mozjpeg and oxipng.

How can I extend the default zune encoders with those libraries? Ideally, I would like to create `zune-mozjpeg` or `zune-oxipng` crates to integrate these libraries with `zune-image`. Can I do that, or would it be better to create internal bindings? What approach would be more suitable?

**Comment by etemesi254** [comment]
Hi, I actually never thought of extending it this way

I don't have the capability to manage things like other encoders and decoders.

Encoders are implemented as traits so anything that can implement `EncoderTrait` is an encoder. A way may be extending ImageFormat enum with `CustomEncoder(Box<dyn EncoderTrait>)` then use save_with and pass the customer ebcoder enum with the encoder nested.

Lemme see if i can come up with a prototype.

**Reply by vlad-salone** [reply]
Maybe add to `ImageFormat` custom field with trait that implements all required functions to get decoder, encoder and file format. Also add to `Image` struct `write_with_encoder` method that can accept any encoder, like `image` crate do.

**Reply by etemesi254** [reply]
yea, that is also possible. Will look into it
